### PR TITLE
Bring missing changes to DataReader and DataWriter

### PR DESCRIPTION
**Pull request by MiguelCompany**
This PR bring changes done to `fastrtps::Subscriber` and `fastrtps::Publisher` into `fastdds::DataReader` and `fastdds::DataWriter`.

Changes are from PRs #968, #1029, #1097 and #1104

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1712/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1473/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1704/)

File: src/cpp/fastdds/publisher/DataWriterImpl.cpp [hunk 1]
```diff
@@ -666,40 +657,41 @@ bool DataWriterImpl::lifespan_expired()
     std::unique_lock<RecursiveTimedMutex> lock(writer_->getMutex());
 
     CacheChange_t* earliest_change;
-    if (!history_.get_earliest_change(&earliest_change))
+    while (!history_.get_earliest_change(&earliest_change))
```
[the remark this answers is not in the document]
Comment (richiware) [reply src/cpp/fastdds/publisher/DataWriterImpl.cpp:660]: I've checked in `fastrtps_deprecated` code and this line is wrong.
```suggestion
    while (history_.get_earliest_change(&earliest_change))
```

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 2]
```diff
@@ -468,40 +474,41 @@ bool DataReaderImpl::lifespan_expired()
     std::unique_lock<RecursiveTimedMutex> lock(reader_->getMutex());
 
     CacheChange_t* earliest_change;
-    if (!history_.get_earliest_change(&earliest_change))
+    while (!history_.get_earliest_change(&earliest_change))
```
Comment (richiware) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp:477]:
```suggestion
    while (history_.get_earliest_change(&earliest_change))
```

Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl.cpp:477]: 👍

File: src/cpp/fastdds/publisher/DataWriterImpl.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/publisher/DataWriterImpl.cpp:660]: 👍

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1732/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1489/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1724/)

**Pull request merged**